### PR TITLE
Perf improvements for effective visibility calculating

### PR DESCRIPTION
**Pull request by Bryanskiy**
related to https://github.com/rust-lang/rust/pull/102026
r? @petrochenkov

### CODE REVIEW

**Comment by petrochenkov** [comment]
If the update operation is a hotspot you need to get rid of all the helper methods it calls and reimplemented it from scratch as a single operation.

What I don't like right now:
- Two nested loops over levels, the logic should go over the previous and parent effective visibilities once and do the minimal number of comparisons.
- `EffectiveVisibility::update` is only called once from `AccessLevels::update` and can be inlined/removed.
- Double search in the map `self.get_effective_vis(id)` + `self.map.insert(id, ...)`, you could use `self.map.entry(...)` instead.

It may also be useful to go back to using `Option<Visibility>`es in `EffectiveVisibility`, in that case fully private visibilities (`None`) can be compared without `is_at_least`.
All of this needs benchmarking though.

Also, as I said in https://github.com/rust-lang/rust/pull/102026#issuecomment-1280352768 it's possible that `update` itself is not the main reason for the slowdown, it's possible that the reason is that we call it for items for which we shouldn't call it at all.
@rustbot author

File: compiler/rustc_resolve/src/access_levels.rs [hunk 1]
```diff
@@ -64,12 +63,11 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
                         self.update(
                             id,
                             binding.vis.expect_local(),
-                            prev_parent_id,
+                            module_id,
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:66]: I don't think this change is correct.
Eff visibility of an import is indeed limited by the previous import in the chain, not the original module.

**Comment by petrochenkov** [comment]
I cannot start a perf run until this PR is rebased.
@rustbot author

File: compiler/rustc_resolve/src/access_levels.rs [hunk 2]
```diff
@@ -96,9 +96,15 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
         parent_id: LocalDefId,
         tag: AccessLevel,
     ) {
+        let module_id = self
+            .r
+            .get_nearest_non_block_module(def_id.to_def_id())
+            .nearest_parent_mod()
+            .expect_local();
+        if nominal_vis == Visibility::Restricted(module_id) {
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:104]:
```suggestion
        if nominal_vis == Visibility::Restricted(module_id) || self.r.visibilities[&parent_id] == Visibilities::Restricted(module_id) {
```
Both `nominal_vis` and `parent_id` set an upper bound on effective visibility, so this condition can be extended like this.

File: compiler/rustc_resolve/src/access_levels.rs [hunk 3]
```diff
@@ -96,9 +96,15 @@ impl<'r, 'a> AccessLevelsVisitor<'r, 'a> {
...
+            return;
+        }
         let mut access_levels = std::mem::take(&mut self.r.access_levels);
-        let module_id =
-            self.r.get_nearest_non_block_module(def_id.to_def_id()).def_id().expect_local();
         let res = access_levels.update(
```
Comment (petrochenkov) [inline compiler/rustc_resolve/src/access_levels.rs:108]:
```suggestion
        self.changed |= access_levels.update(
```
Nit: unnecessary variable.

File: compiler/rustc_privacy/src/lib.rs [hunk 4]
```diff
@@ -922,29 +922,32 @@ pub struct TestReachabilityVisitor<'tcx, 'a> {
...
-                            if restricted_id.is_top_level_module() {
-                                "pub(crate)".to_string()
-                            } else if *restricted_id == self.tcx.parent_module_from_def_id(def_id) {
-                                "pub(self)".to_string()
-                            } else {
-                                format!("pub({})", self.tcx.item_name(restricted_id.to_def_id()))
-                            }
+            let module_id = self.tcx.parent_module_from_def_id(def_id);
+            let effective_vis =
+                self.access_levels.get_effective_vis(def_id).copied().unwrap_or(
```
Comment (petrochenkov) [inline compiler/rustc_privacy/src/lib.rs:927]:
```suggestion
                self.access_levels.get_effective_vis(def_id).copied().unwrap_or_else(||
```

**Comment by petrochenkov** [comment]
r=me after squashing commits.
@rustbot author

**Review by ghost — APPROVED** [review]

**Pull request merged**